### DISCUSSION TITLE
Why Microsoft.Extensions.ServiceDiscovery P5 version removing several algorithms for load balancing, leaving only RoundRobinServiceEndpointSelector

### DISCUSSION THREAD

**Discussion by vipwan** [microsoft/aspire#3640] [Q&A] [answered]
Why P5 version removing several algorithms for load balancing, leaving only RoundRobinServiceEndpointSelector, if so, why is the IServiceEndpointSelector interface still internal, if I need a custom load balancing algorithm, I won't be able to start

```c#
internal interface IServiceEndpointSelector {}
```

Whether the `IServiceEndpointSelector` interface can be marked as public

**Comment by davidfowl** [comment]
cc @ReubenBond

**Reply by ReubenBond** [reply, marked as answer]
We opted to remove the interfaces for now and wait until we found a stronger need to expose them. If there is a request from the community, we could make them public again and bring back some of the implementations (RoundRobin, PickFirst). If it's something you are interested in, please open an issue. The PowerOf2Choices implementation never worked as advertised because it relies on feedback from the endpoints and we don't have any defined way to get that feedback in HttpClient (backend servers need to report their load for us)

**Reply by ROMYIM** [reply]
`RoundRobinServiceEndpointSelector` is  too simple to select a true endpoint.
For example:
A service instance usually linstens multipart ports for many schemes, such as http, https, and so on.
Not only http  but also grpc request was sent to service B by service A.
My custom `IServiceEndpointProvider` provides many service endpoints with multipart schemes by the same service name, but the `RoundRobinServiceEndpointSelector` can not select a true endpoint according to the URI scheme.
It is nessary to make IServiceEndpointSelector interface public

**Reply by rem-coding** [reply]
I'm surprised not more feedback on this item. Only supporting round robin is very limiting. It would be great to bring back PickFirst, and allow for custom providers.

**Reply by chuanboz** [reply]
The issue is opened in https://github.com/dotnet/aspire/issues/4890, please vote for this issue if you also ask for it.

**Comment by vipwan** [comment]
Considering that the `IServiceEndPointProvider ` of other three parties may be connected in the future, special factors such as weight, validity period and so on may be involved
